# Integrating heterogeneous data sources to predict gene essentiality in Saccharomyces cerevisiae
Course project for CU Boulder CSCI 5622: Machine Learning with Dr. Jordan Boyd-Graber

**Team members**: Nicolas Metts, Matthew Pennington, Rani Schwindt and Carter Tillquist

### (Quick intro for group members, will be formalized for the write-up)
An "essential" gene is one which, when absent/deleted, confers a lethal phenotype (the yeast dies). We're proposing that gene essentiality can be predicted using a weighted combination of features, like different aspects of a geneâ€™s sequence or its protein product to find the pattern that makes a gene essential. There are lots of motivating reasons to perform an analysis like this. Here we're using all genes from *Saccharomyces cerevisiae*, which is very well studied and characterized, but other yeast species are not well characterized at all.  Therefore, a pattern that helps us determine gene essentiality in yeast can be applied to a related organism in order to get an idea of its essential genes. Also, [many yeast genes have homologs in the human](http://downloads.yeastgenome.org/curation/literature/functional_complementation.tab).

### Features we currently have for all genes:
* Localization: nucleus, mitochondria, cytoplasm, ER, vacuole or other
* Chromosome
* Chromosome position
* Number of transmembrane helices
* Length of protein product measured in amino acids
* Hydrophobicity or hydrophilicity
* [Effective number of codons](http://en.wikipedia.org/wiki/Effective_number_of_codons)
* [Codon adaptation index](http://en.wikipedia.org/wiki/Codon_Adaptation_Index)
* GC content
* Proportion of codons one third-base away from a stop codon
* Proportion of rare amino acids in translated ORF
* Number of [BLAST hits](https://en.wikipedia.org/wiki/BLAST) in prokaryotes
* Number of BLAST hits in all yeast species
* Number of BLAST hits in 6 close yeast species
* Number of protein interaction partners

|       Feature        | Current format |     Transformed format?     |
|----------------------|----------------|-----------------------------|
| Localization         | Text or binary | 'nucleus', 'cytoplasm' etc  |
|       Chromosome     |      Integer   |    'Chr1', 'Chr2'           |
| Chromosome position  |     Integer    |     'Low', 'Med', 'High'    |
| Length aa            |     Integer    |   'Short', 'Med', 'Long'    |
|  TM helices          |     Integer    | 'LowTM', 'MedTM', 'HighTM'  |
| Hydrophobicity       |    -inf to inf | 'Hydrophobic', 'Hydrophilic'|
| Number codons        |     Integer    |                             |   
| Codon adaptation ind |    0 to 1      |                             |
|    GC content        |     0 to 1     |                             |
| BLAST hits in yeast  |      Integer   |                             | 
| BLAST hits in 6 yeast|      Integer   |                             | 
|BLAST hits in proks   |      Integer   |                             |
| intxn_partners       |      Integer   |                             |


### Previous baseline

A [paper by Seringhaus et al](http://genome.cshlp.org/content/16/9/1126.long) compiled 3 datasets: Wysocki (1999), Vandenbol (2000) and Stevenson (2001) and used a Bayesian framework to predict essential genes in *S. cerevisiae* and *S. miktae* (another type of yeast). 

**From their abstract:**

"Essential genes are required for an organism's viability, and the ability to identify these genes in pathogens is crucial to directed drug development. Predicting essential genes through computational methods is appealing because it circumvents expensive and difficult experimental screens. Most such prediction is based on homology mapping to experimentally verified essential genes in model organisms. We present here a different approach, one that relies exclusively on sequence features of a gene to estimate essentiality and offers a promising way to identify essential genes in unstudied or uncultured organisms. We identified 14 characteristic sequence features potentially associated with essentiality, such as localization signals, codon adaptation, GC content, and overall hydrophobicity. Using the well-characterized baker's yeast Saccharomyces cerevisiae, we employed a simple Bayesian framework to measure the correlation of each of these features with essentiality. We then employed the 14 features to learn the parameters of a machine learning classifier capable of predicting essential genes. We trained our classifier on known essential genes in *S. cerevisiae* and applied it to the closely related and relatively unstudied yeast Saccharomyces mikatae. We assessed predictive success in two ways: First, we compared all of our predictions with those generated by homology mapping between these two species. Second, we verified a subset of our predictions with eight in vivo knockouts in *S. mikatae*, and we present here the first experimentally confirmed essential genes in this species." -Seringhaus et al. Genome Research (2002)

### TO DO:
* Figure out how to best represent features
* Feature engineering
* Pick machine learning algorithm (based on feature representation chosen)


